tests: drivers: uart: uart_async_api: Rework to allow testing of multiple instances

nordic-krch opened the pull request:
Test is using a single instance but Nordic has SoCs which has UART instances that are a bit different and we would like to test multiple of them in the test.

Reworked the test to have an array of DUT instances with only one currently added.

Due to userspace setup test is using the approach where each test is in separate suite so it is not that obvious how to may it mutli-instance and maybe there is a better way than I took.

nordic-krch committed:
tests: drivers: uart: async_api: Add uart120 instance to nrf54h20dk

Add second instance to be tested on nrf54h20dk. uart120 is a fast UARTE
which works on fixed pin locations. It is not available for cpuppr core.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_api: Rework for multi instance

Rework the test to be able to run on multiple instances.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: async_api: Add missing static keyword

Multiple variable in the test were missing static keyword.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

bjarki-andreasen commented:
We can also add additional overlays for each UART instance we want to test, and keep the test suite simple, building and running for just one UART. This way, for boards which require a physical connection between RX and TX can reuse the same pins for each UART instance.

nordic-krch commented:
@bjarki-andreasen yes we can keep approach with dedicated overlay and configuration in `testcase.yaml`, but then we are spoiling `testcase.yaml` and increasing unique files (overlays) that must be maintained. This approach requires also multiple tests to be build&run on same target (impact CI time). We already took that approach here: https://github.com/zephyrproject-rtos/zephyr/blob/ac52bd629d705811858c8bb38ca7a43e4e81182b/tests/drivers/spi/spi_loopback/testcase.yaml#L181 where test for special SPIM120 (fast) was added. A configuration was added for `nrf54h20dk/nrf54h20/cpuapp`. I don't like this because we would like to test that also on other cores in nrf54h20, nrf54l15 also has fast SPI, maybe more cores in the future will need that. We can end up with dozens of custom overlays and configurations. With multi-instance approach you can still do single-instance + custom overlay if you want to reuse pins. In that particular case (fast peripherals on nrf54h) pins cannot be reused multi-instance approach can be used.

bjarki-andreasen commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -24,40 +24,50 @@ K_SEM_DEFINE(rx_buf_coherency, 0, 255);
+	const char *name;
+};
 
+static ZTEST_DMEM struct dut_data duts[] = {
+	{
+		.dev = DEVICE_DT_GET(UART_NODE),
+		.name = DT_NODE_FULL_NAME(UART_NODE),
+	},
+	/* More instances can be added here. */
+};
```
Lets expand this to automatically find additional devices in the overlays: dut0, dut1, dut2 ... :)
```
static ZTEST_DMEM struct dut_data duts[] = {
	{
		.dev = DEVICE_DT_GET(UART_NODE),
		.name = DT_NODE_FULL_NAME(UART_NODE),
	},
#if DT_NODE_EXISTS(DT_NODELABEL(dut1))
	{
		.dev = DEVICE_DT_GET(DT_NODELABEL(dut1)),
		.name = DT_NODE_FULL_NAME(DT_NODELABEL(dut1)),
	},
#endif
#if DT_NODE_EXISTS(DT_NODELABEL(dut2))
	{
		.dev = DEVICE_DT_GET(DT_NODELABEL(dut2)),
		.name = DT_NODE_FULL_NAME(DT_NODELABEL(dut2)),
	},
#endif
        ...
};
```

bjarki-andreasen reviewed (changes requested):
Please add the suggested change and update at least one of the overlays to designate more than one `dut` to validate that it is working, and to set an example :)

bjarki-andreasen commented:
> @bjarki-andreasen I will be on 2 weeks vacation and I'll get back to that after that. I've added it here https://github.com/zephyrproject-rtos/zephyr/pull/75462/commits/8d9b93bbcaab03acbcfb24dedea6bceb9784e989 which is a draft PR.

ok, enjoy your vacation :)

nordic-krch committed:
drivers: serial: nrfx_uarte: Fix pin retention

b6d45423c60b Added support for pin retention but it was added
only to the case when device PM is used. There is another mode
in which UARTE is disabled when idle (low power mode) and in
that case pin retention must also be added.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
I've added an example for using multi-instance test. For nrf54h20dk//cpuapp `uart137` and `uart120` are tested.
PR contains one commit from #81273 which is needed to make uart120 work without device runtime PM.

bjarki-andreasen reviewed (approved):
Looks nice!
